R/bootstrap.R

```diff
@@ -1695,90 +1695,342 @@ verbatimTextOutput <- function(outputId) {
-                       hoverDelayType = c("debounce", "throttle"), inline = FALSE) {
-  if (is.null(clickId) && is.null(hoverId)) {
-    hoverDelay <- NULL
-    hoverDelayType <- NULL
-  } else {
-    hoverDelayType <- match.arg(hoverDelayType)[[1]]
+imageOutput <- function(outputId, width = "100%", height="400px",
+                        click = NULL, dblclick = NULL,
+                        hover = NULL, hoverDelay = NULL, hoverDelayType = NULL,
+                        brush = NULL,
+                        clickId = NULL, hoverId = NULL,
+                        inline = FALSE) {
+
+  if (!is.null(clickId)) {
+    shinyDeprecated(
+      msg = paste("The 'clickId' argument is deprecated. ",
+                  "Please use 'click' instead. ",
+                  "See ?imageOutput or ?plotOutput for more information."),
+      version = "0.11.1"
+    )
+    click <- clickId
+  }
+
+  if (!is.null(hoverId)) {
+    shinyDeprecated(
+      msg = paste("The 'hoverId' argument is deprecated. ",
+                  "Please use 'hover' instead. ",
+                  "See ?imageOutput or ?plotOutput for more information."),
+      version = "0.11.1"
+    )
+    hover <- hoverId
+  }
+
+  if (!is.null(hoverDelay) || !is.null(hoverDelayType)) {
+    shinyDeprecated(
+      msg = paste("The 'hoverDelay'and 'hoverDelayType' arguments are deprecated. ",
+                  "Please use 'hoverOpts' instead. ",
+                  "See ?imageOutput or ?plotOutput for more information."),
+      version = "0.11.1"
+    )
+    hover <- hoverOpts(id = hover, delay = hoverDelay, delayType = hoverDelayType)
   }
 
   style <- if (!inline) {
     paste("width:", validateCssUnit(width), ";", "height:", validateCssUnit(height))
   }
 
+
+  # Build up arguments for call to div() or span()
+  args <- list(
+    id = outputId,
+    class = "shiny-image-output",
+    style = style
+  )
+
+  if (!is.null(click)) {
+    # If click is a string, turn it into clickOpts object
+    if (is.character(click)) {
+      click <- clickOpts(id = click)
+    }
+
+    args <- c(args, list(
+      `data-click-id` = click$id,
+      `data-click-clip` = click$clip
+    ))
+  }
+
+  if (!is.null(dblclick)) {
+    if (is.character(dblclick)) {
+      dblclick <- clickOpts(id = dblclick)
+    }
+
+    args <- c(args, list(
+      `data-dblclick-id` = dblclick$id,
+      `data-dblclick-clip` = dblclick$clip,
+      `data-dblclick-delay` = dblclick$delay
+    ))
+  }
+
+  if (!is.null(hover)) {
+    if (is.character(hover)) {
+      hover <- hoverOpts(id = hover)
+    }
+
+    args <- c(args, list(
+      `data-hover-id` = hover$id,
+      `data-hover-clip` = hover$clip,
+      `data-hover-delay` = hover$delay,
+      `data-hover-delay-type` = hover$delayType
+    ))
+  }
+
+  if (!is.null(brush)) {
+    if (is.character(brush)) {
+      brush <- brushOpts(id = brush)
+    }
+
+    args <- c(args, list(
+      `data-brush-id` = brush$id,
```

Comment:
I don't particularly have a preference of what the implementation is. My only two desires here are 1) less code, 2) make functions that strongly couple with each other colocated in the source code.